README.md
# harvestR <img src="man/figures/logo.png" align="right" height="125" width="125" />

harvestR is a get wrapper to extract tables from the Harvest API v2. 

## Installation
To install harvestR run the following:
``` r
devtools::install_github("propellerpdx/harvestR")
```

## Credentials
Harvest requires credentials to make any API requests. Harvest supports personal access tokens and oauth2, for more info refer to [Harvest](https://help.getharvest.com/api-v2/authentication-api/authentication/authentication/).

### PAT  
To authenticate using a personal access token, you need an account ID and personal access token which can be obtained [here](https://id.getharvest.com/developers). The user account is shared by all users in that account and personal access token is specific to the user. These credentials can be used in `get_table()` by passing them to the `user` and `key` parameters. 
``` r
projects <- get_table(table = 'projects', user = 'not_a_real_account', key 'Bearer fake_personal_access_token')
```
**Note:** you must add `Bearer ` including the space to your personal access token in order for it to work. This is a [Harvest requirement](https://help.getharvest.com/api-v2/authentication-api/authentication/authentication/).

harvestR provides functions for users to manage their credentials using the [keyring](https://github.com/r-lib/keyring) package. This allows users to avoid storing secrets in their code and allows them to keep from retyping them every time they make a call. To do so run the below code and add the account ID and personal access tokens in the prompts:
``` r
create_harvest_creds()
```  
**Note:** you must add `Bearer ` including the space to your personal access token in order for it to work. This is a [Harvest requirement](https://help.getharvest.com/api-v2/authentication-api/authentication/authentication/).

Once the credentials are created, they are saved on the system keyring so they can be accessed in future interactive R sessions. By running the below code harvestR will automatically find your credentials when running `get_table()`.  

``` r
set_harvest_creds()
```  

If at any point you need to update your token, you can call `create_harvest_creds()` again - it will delete the old credentials once you have provided the new ones. If you want to delete the credentials from your keyring and not add new ones, run `harvestR::delete_harvest_creds()`.

### Oauth  
If users require oath support, tokens can passed `get_table()` as a parameter `token`. Tokens are forwarded to `httr::config`. This functionality has not been tested as of this release.
``` r
get_table(table = 'projects', token = my_token)
```  
  
### Non-Interactive Sessions  
For non-interactive users such as Rscript, harvestR will still attempt to retrieve your account ID and personal access token from the environment variables `HARVEST_ACCOUNT_ID` and `HARVEST_PAT`. For additional information for storing credentials refer to [usethis](https://usethis.r-lib.org/articles/articles/git-credentials.html).
  
## Get Tables
The `get_table()` function gets tables from the [Harvest API](https://help.getharvest.com/api-v2/) by name and coerces the data into a data frame. A number of optional parameters are in place to add query parameters, configure parallel processing, and provide messaging for logs. A few examples are provided below.

``` r
# Request the projects table
projects <- get_table(table = 'projects')

# Request the time_entries table for 3 months
time_entries <- get_table(table = 'time_entries', query = list(from = '2018-01-01', to = '2018-03-31'))

# Request the time_entries table for 3 months in with parallel requests
time_entries <- get_table(table = 'time_entries', query = list(from = '2018-01-01', to = '2018-03-31'), plan_options = list(strategy = "multicore"))

# Request the time_entries table for 3 months in parallel mode with automatic retries lowered to 1 (times) and messaging on (quiet)
# Windows does not support multicore plans, users must use multisession
time_entries <- get_table(table = 'time_entries', query = list(from = '2018-01-01', to = '2018-03-31'), plan_options = list(strategy = "multisession"), times = 1, quiet = F)

# Request using an httr option (see `httr::httr_options()`) and an oauth token
projects <- get_table(table = 'projects', http_version = 2, token = my_token)
```


news.md
# harvestR 1.1.0

## General features

* Added messaging to `get_table()` to address verbose output issue.
* The strategy, auto_retry, and verbose parameters are deprecated and replaced by plan_options, times, and quiet, respectively.
* Added credential management workflow, special thanks to [joethorley](https://github.com/joethorley) for suggesting the idea - sorry for the long wait!

## Messaging
The `get_table()` function had a `verbose` parameter was passed to `httr::verbose()`. Implementing parallel requests in `get_table()` broke the verbose output. The future package has a [known limitation](https://cran.r-project.org/web/packages/future/vignettes/future-2-output.html) with messaging across parallel threads. We expect many users will want to take advantage of parallel requests so we added custom messaging inside the `get_table()` which provides information for logging. The messaging can be turned on by setting `quiet = FALSE` in the `get_table()` call. To divert output to a log file you can wrap calls with logging functions. 

## Parameter changes
* The `verbose` parameter is deprecated because it does not work on parallel threads (refer to Messaging above for more details). The `quiet` parameter is the suggested replacement. Verbose will still be passed to `httr::config()`, but a warning will be issued for the time being to make sure users are aware of the behavior changes. 
* The `strategy` parameter is deprecated in favor of plan_options. The `strategy` parameter was passed to `future::plan()`, but did not allow users to modify any other parameters inside `future::plan()`. The new `plan_options` parameter allows users to pass any parameters into `future::plan()` in a named list.  
* The `auto_retry` parameter is deprecated because it was used to retry get requests that failed in custom logic. We moved to the `httr::RETRY()` function in order to simplify. The `times` parameter is the suggested replacement. 
* Additional arguments in ... were specifically forwarded to httr functions, but only a few were supported. `get_table()` now forwards listed arguments to `httr::RETRY` and attempts to forward all other arguments to `httr::config()` including token (for oath requests). 

## Credential Management
Joe Thorley made the suggestion of adding credential management functions to make fetching keys easier within the library. We added a few functions that are intended to provide a credential management workflow for harvestR. They enable interactive users to securely save credentials on their system using the [keyring](https://github.com/r-lib/keyring) package. Scheduled jobs can still take advantage of the system with system environment variables. Refer to the harvestR package readme for more info or refer to the `create_harvest_creds()` function to get started. 

# harvestR 1.0.0

## General features

* `get_table()` gains new arguments to allow auto retries, parallel get requests, and more verbose messaging.

## Executing get_table requests in parallel

* The `strategy` parameter is passed to `future::plan()` which sets a plan for how future executions are resolved. The parameter does not provide full access to all `future::plan()` parameters and functionality because the current use cases do not call for it. 

## Automatic retries

* The `auto_retry` parameter will automatically retry get requests that fail (i.e. non-200 responses) up to 3 times. It does not have intelligence only perform retries depending on the response code. It is primarily useful if the user is running batches over a poor internet connection. 


R/get_harvestR.r
#' @title Harvest API v2 get request wrapper
#'
#' @description Submits get requests to the Harvest API v2. The function is a
#'   wrapper around the httr::GET function and returns the Harvest table.
#'
#' @param table Harvest API table that does not require specific ID's for the
#'   GET request. Refer to \href{https://help.getharvest.com/api-v2}{Harvest
#'   APIv2}. Possible values include the following as a string:
#'   (\code{"clients"} or \code{"invoices"} or \code{"estimates"} or
#'   \code{"expenses"} or \code{"tasks"} or \code{"time_entries"} or
#'   \code{"projects"} or \code{"roles"} or \code{"users"})
#' @param user Harvest account number for authentication. Register at
#'   \href{https://id.getharvest.com/developers}{Harvest Developers}.
#' @param key Harvest API key for authentication. Register at
#'   \href{https://id.getharvest.com/developers}{Harvest Developers}.
#' @param email Optional argument, requester email address for the "From" header
#'   in the get request.
#' @param query Optional argument, API query parameters to be provided in a
#'   list. Refer to \href{https://help.getharvest.com/api-v2}{Harvest APIv2} for
#'   acceptable parameters for each table. A few examples include:  (\code{from
#'   = "2018-1-1"} or \code{to = "2018-3-31"} or \code{project_id = "1234"} or
#'   \code{client_id = "1234"} or \code{user_id = "1234"})
#' @param verbose logical; passed to httr; A verbose connection provides much
#'   more information about the flow of information between the client and
#'   server. \href{https://github.com/r-lib/httr}{httr documentation}
#'
#' @return tbl_df
#'
#' @examples
#'\dontrun{
#' api_key <- 'your_user_id'
#' get_table(table = 'projects',
#'           user = user_id,
#'           key = api_key)
#' get_table(table = 'time_entries',
#'           user = user_id,
#'           key = api_key,
#'           email = 'your_email',
#'           query = list(from = '2018-01-01',
#'                        to = '2018-03-31')
#'           )
#'}
#'
#' @author Mark Druffel, \email{mdruffel@propellerpdx.com}
#' @references \url{https://id.getharvest.com/developers},
#'   \url{https://help.getharvest.com/api-v2},
#'   \url{https://github.com/r-lib/httr}
#'
#' @importFrom magrittr %>%
#' @import httr
#'
#' @export
get_table <- function(table = NULL,
                      user = harvest_id(),
                      key = harvest_pat(),
                      email = '',
                      query = NULL,
                      verbose = FALSE) {

  if(!is.null(key) && !grepl("^Bearer ", key))
    key <- paste0('Bearer ', key)

  if(!is.null(query$return_df)){
    return_df <- query[[length(query)]]
    query <- query[[1:(length(query)-1)]]}

  if(is.null(query$page)){query$page <- '1'}

  url <- paste0('v2/',table) %>% httr::modify_url(url="https://api.harvestapp.com",path=.,query=query)

  # Harvest's API responds that it can support HTTP/2, but it doesn't seem to
  # actually be able to. Some versions of curl use HTTP/2 and fail as a
  # consequence while others figure it out and get by it, using HTTP/1.1 (e.g.
  # the curl version installed on AWS EC2 / Lambda suffers from this problem)
  # http_version=2 forces HTTP/1.1
  # Full http_version options can be seen here
  # https://github.com/curl/curl/blob/master/include/curl/curl.h by searching
  # for "http_version"
  # Full httr docs https://cran.r-project.org/web/packages/httr/httr.pdf

  response_df <- url %>%
    purrr::map(.,
               function(x)
                 httr::with_config(
                   config = config(verbose = verbose,
                                   http_version = 2),
                   httr::GET(
                     x,
                     httr::add_headers(
                       "Harvest-Account-ID" = user,
                       Authorization = key,
                       'User-Agent=Propeller R API Helper (mdruffel@propellerpdx.com)',
                       'From' = email
                     ),
                     query = query
                   )
                 )) %>%
    purrr::map(., function(x) httr::content(x, as="text", encoding = "UTF-8")) %>%
    purrr::map(., function(x) jsonlite::fromJSON(x, flatten = T))

  next_page <- response_df %>% purrr::map('next_page')

  if(!exists('return_df')){return_df <- NULL}
  return_df <- response_df %>%
    purrr::map(.,paste0(table)) %>%
    purrr::flatten_df(.) %>%
    dplyr::bind_rows(.,return_df)

  query$page <- paste0(unlist(next_page))
  if(is.null(unlist(next_page))){return(return_df)}else{
    query <- list(query=query,return_df=return_df)

    if(response_df %>% purrr::map_dbl(.,'total_pages')>99){naptime::naptime(.1)}

    get_table(table=table,user=user,key=key,email=email,query=query)}
}
